### PR TITLE
Better character cell width with FreeType

### LINKED ISSUE #1020: Wrong font width (maximum advance freetype2 value)
debian Sid, x11

`maximum advance` via freetype2 is too wide for some fonts.

Same issue as #902 and #990 and I'm still unable to solve this.
I've tried the [workaround](https://gist.github.com/mikoim/498dfcdd39439316e3eb9b8897914302) proposed but the patch is unapplicable. I've reproduced the same modifications and applied [this](https://gist.github.com/avivace/9d3fddcb43e853d53cadbca4ed11d27e) patch. Still nothing. 

Narrow fonts like Iosevka or Pragmata Pro have this issue:
image

I'm currently just setting offset x to -8.0 and it looks pretty good, almost identical to my other terminal emulator.

Offsetting by `-8.0` however, causes a strange behaviour on the dimensions column  value (I have to set `1000` to see ~80 colums)

Still, what's the status of this?

### PR DESCRIPTION
**Pull request by rkanati**
This should fix #1020, #710, and #902 

I'm happy to rework this if there are any problems with it.

### CODE REVIEW

**Comment by maximbaz** [comment]
Nice job. Personally I was always able to reproduce this issue only with non-monospace fonts (such as `Ubuntu` contrary to `Ubuntu Mono`, or `DejaVu Sans` contrary to `DejaVu Sans Mono`), with this PR I confirm that:

1. Monospace fonts are still rendered correctly, identically to master
2. Non-monospace fonts look not perfect, but much better indeed

Picture shows it all, font name is written in the prompt line:

image

**Comment by chrisduerr** [comment]
Non-monospace fonts shouldn't work. They simply can't be fitted into a grid like a terminal emulator.

However what alacritty has done until now is that it takes all characters in a font and checks their width, then uses an average of that as the cell width.

And with some fonts that has caused some issues apparently because they contained large characters that changed this average significantly and affected the overall cell width.

This branch seems like a good solution to me because the size of normal ascii characters like `0` is always the exact size of a cell in a monospace font as far as I know. So this should be a sufficient solution for this problem.

Please do correct me if I'm wrong though.

**Comment by rkanati** [comment]
Well, the problem was that, with the FreeType driver, the cell width _wasn't_ coming from average advance, it was coming from _max_ advance. This could be badly wrong even on an otherwise well-behaved mono font, if it happened to contain CJK stuff of fancy presentation forms.

Just using '0' seems like it should be fine - it's what the CoreText driver does, and none of the referenced bugs are on Mac. If it needs to be more robust, averaging the widths of all of Latin-1 should be pretty solid.

**Comment by chrisduerr** [comment]
In a monospaced font there should be no difference between latin1 character widths. So this should be fine.

However if you would want to check all latin1 characters, it should probably check the max advance and not the average.

File: font/src/ft/mod.rs [hunk 1]
```diff
@@ -225,13 +243,14 @@ impl FreeTypeRasterizer {
...
+            let load_flags = Self::ft_load_flags(pattern);
             let face = Face {
                 ft_face: ft_face,
                 key: FontKey::next(),
-                load_flags: Self::ft_load_flags(pattern),
+                load_flags: load_flags,
                 render_mode: Self::ft_render_mode(pattern),
                 lcd_filter: Self::ft_lcd_filter(pattern),
-                non_scalable: non_scalable,
+                non_scalable: non_scalable
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Tiny nitpick but these changes from line 246 to line 253 are not necessary, right? At least I can't see why this got changed.

**Comment by jwilm** [comment]
> monospaced font there should be no difference between latin1 character widths

Unfortunately, this isn't quite true. "Monospaced fonts" are more convention than anything else, and it's only ever really enforced at the application level such as a terminal where there are hard lines and columns.

> However if you would want to check all latin1 characters, it should probably check the max advance and not the average.

I think back in a very early version of Alacritty it was implemented using max advance, and that had some issues which lead us to using freetype metrics directly. I think the average might actually be the most reasonable approach here.

File: font/src/ft/mod.rs (hunk 1 above)
Comment (rkanati) [reply font/src/ft/mod.rs]: Yeah, those were just screw-ups I forgot to clean up. Fixing.

**Comment by rkanati** [comment]
So those little whoopsies are fixed now.

@chrisduerr If you mean the max of advance _of_ those Latin-1, then yeah, you're right.

@jwilm I'm not sure what you mean about it using max advance in a very early version - misuse of whole-font max advance is exactly what this PR fixes.

**Comment by anxiousmodernman** [comment]
I reported #1043 and built this branch, and it's an improvement, but now the default config yields unevenly-spaced characters, like this:
img_20180121_104725

**Comment by maximbaz** [comment]
@anxiousmodernman could you run the attached binary and post the output of stdout here?

[font-test.zip](https://github.com/jwilm/alacritty/files/1650068/font-test.zip)

**Comment by chrisduerr** [comment]
This looks to me like you're not using a monospace font @anxiousmodernman.

**Comment by anxiousmodernman** [comment]
It crashes on startup. I captured the i3 logs here

https://pastebin.com/jDYvM4QM

**Comment by maximbaz** [comment]
It is supposed to crash, thanks for the log 🙂 All characters have different width, so @chrisduerr is right, you are using a non-monospace font, this is not supported. Try changing the font in the alacritty config.

**Comment by anxiousmodernman** [comment]
So the alacritty config is the generic "monospace". I've never changed it on any machine. I think the difference this time is the bare install. I'm running Cinnamon or KDE desktops on my other machines, and it might provide a font such that I haven't had to configure anything in those envs before. I don't know anything about fonts, so this might take me a while.

**Comment by maximbaz** [comment]
`$ fc-match -s monospace` will show you which fonts are being currently selected, the priority is from top to bottom. The easiest way to is to install a monospace font that you like and put its name here:

https://github.com/jwilm/alacritty/blob/b396a9a75350c5b1c82f5aae06a68dbd035df83c/alacritty.yml#L61-L64

Good luck 🙂

**Comment by jwilm** [comment]
@rkanati thank you for this PR! I'm excited this closes all of those issues at once. 😄 

If you are interested in some follow-up work, I was thinking it might make sense to cache the statistics computed here so they aren't reloaded/calculated every time `full_metrics` is called. I filed #1045 to track this.

**Pull request merged**